epymodelingsuite/filtering.py

```diff
@@ -0,0 +1,120 @@
+    surveillance_location_col: str = "location_iso",
+    surveillance_target_col: str = "hospitalizations",
+    surveillance_date_col: str = "target_end_date",
+    simulation_target: str | list[str] = "hospitalizations",
+) -> tuple[list[dict], pd.DataFrame]:
+    """
+    Filters simulation projections based on their similarity to observed surveillance data.
+
+    For each trajectory in the runner output, calculates a distance to the observed
+    surveillance data at the specified `final_fitting_date`. Returns only the top
+    fraction of trajectories with the smallest distances.
+
+    If `simulation_target` is a list, sums the simulation values at the fitting date
+    for all specified targets.
+
+    Args:
+        runner_output: Object containing simulation results, must have 
+            `results.projections["baseline"]` and `results.projection_parameters["baseline"]`.
+        surveillance_data: DataFrame containing observed data.
+        top_fraction: Float between 0 and 1 specifying fraction of projections to keep.
+        final_fitting_date: Date (string, pd.Timestamp, or datetime) to filter observed data.
+        distance_function: Callable that takes two dicts with key 'data' and returns a float.
+        surveillance_location_col: Column in `surveillance_data` specifying location.
+        surveillance_target_col: Column in `surveillance_data` with observed values.
+        surveillance_date_col: Column in `surveillance_data` with dates.
+        simulation_target: Column name or list of column names in simulation projections to compare.
+
+    Returns:
+        Tuple of:
+            - filtered_projections: List of projections corresponding to the top fraction (in the 
+            same format as they appear in CalibrationOutput). 
+            - filtered_projection_parameters: DataFrame of parameters corresponding to these projections.
+
+    Raises:
+        ValueError: If `top_fraction` is not in (0, 1], or if no data exists for `final_fitting_date`.
+        KeyError: If expected columns are missing from data.
+    """
+    # Validate inputs
+    if not (0 < top_fraction <= 1):
+        raise ValueError(f"`top_fraction` must be in (0, 1], got {top_fraction}")
+
+    if surveillance_location_col not in surveillance_data.columns:
+        raise KeyError(f"Column '{surveillance_location_col}' not found in surveillance_data")
+    if surveillance_target_col not in surveillance_data.columns:
+        raise KeyError(f"Column '{surveillance_target_col}' not found in surveillance_data")
+    if surveillance_date_col not in surveillance_data.columns:
+        raise KeyError(f"Column '{surveillance_date_col}' not found in surveillance_data")
+
+    # Convert final date to Timestamp
+    final_fitting_date = pd.Timestamp(final_fitting_date)
+
+    # Get location and filter data
+    location = convert_location_name_format(runner_output.population, "ISO")
+    surveillance_location = get_data_in_location(
```

Comment:
Could have better name for clarity e.g. `surveillance_data_at_location`. `surveillance_location` sounds like it contains the name of location.